netket/utils/array.py

```diff
@@ -0,0 +1,50 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from dataclasses import dataclass
+
+import numpy as np
+
+from .types import Array, DType
+
+
+@dataclass(frozen=True)
+class HashableArray:
+    """
+    This class wraps a numpy or jax array in order to make it hashable and
+    equality comparable (which is necessary since a well-defined hashable object
+    needs to satisfy :code:`obj1 == obj2` whenever :code:`hash(obj1) == hash(obj2)`.
+
+    The underlying array can also be accessed using :code:`numpy.asarray(self)`.
+    """
+
+    wrapped: Array
+    """The wrapped array. Note that this array is read-only and modifying it without
+    recreating `HashableArray` will result in incorrect behavior."""
+
+    def __post_init__(self):
+        if isinstance(self.wrapped, np.ndarray):
+            self.wrapped.flags.writeable = False
```

Comment:
This will freeze the array also outside...
We should maybe copy it?
```python
a = np.arange(10)
b = nk.utils.HashableArray(a)
a[1] = 10 #error
```

Reply:
> This will freeze the array also outside...

Yes, this was intended because if it is modified outside the hash breaks.
We could always copy it I guess...

Reply:
yeah the semantically correct thing to do would be a copy

Reply:
Fair enough. I'll make a copy.